app/src/org/commcare/android/adapters/HomeScreenAdapter.java

```diff
@@ -0,0 +1,186 @@
+        setOnClickListenerForButton(buttonIndex, listener);
+    }
+
+    public SquareButtonWithNotification getButton(int androidCode, boolean lookupID){
+        return buttons[getButtonIndex(androidCode, lookupID)];
+    }
+
+    public void setNotificationTextForButton(int androidCode, boolean lookupID, String notificationText) {
+        SquareButtonWithNotification button = getButton(androidCode, lookupID);
+        if (button != null) {
+            button.setNotificationText(notificationText);
+            notifyDataSetChanged();
+        }
+    }
+
+    @Override
+    public int getCount() {
+//        return buttonsResources.length;
+        return visibleButtons == null ? buttonsResources.length : visibleButtons.size();
+    }
+
+    @Override
+    public Object getItem(int position) {
+        return buttons[position];
+    }
+
+    @Override
+    public long getItemId(int position) {
+        return buttonsResources[position];
+    }
+
+    @Override
+    public View getView(int position, View convertView, ViewGroup parent) {
+        View view;
+        if(!isInitialized){
+            visibleButtons = new LinkedList<SquareButtonWithNotification>();
+            Log.i("HomeScrnAdpt","Creating all buttons because got a null in position " + position);
+            for (int i = 0; i < buttons.length; i++) {
+                if (buttons[i] != null) continue;
+                SquareButtonWithNotification button = (SquareButtonWithNotification) LayoutInflater.from(context)
+                        .inflate(buttonsResources[i], parent, false);
+                buttons[i] = button;
+                Log.i("HomeScrnAdpt","Added button " + button + "to position " + i);
+
+                View.OnClickListener listener = buttonListeners[i];
+                // creating now, but set a clickListener before, so we'll add it to this button...
+                if(listener != null) {
+                    button.setOnClickListener(listener);
+                    Log.i("HomeScrnAdpt","Added onClickListener " + listener + " to button in position " + i);
+                }
+                if( i == position ) view = button;
+                if(!hiddenButtons[i]) visibleButtons.add(button);
+            }
+            isInitialized = true;
+        }
+        if(position < 0 || position >= getCount()) return null;
+        if(convertView != null) {
+            return convertView;
+        } else {
+            SquareButtonWithNotification btn = visibleButtons.get(position);
+
+            if(btn == null) {
+                Log.i("HomeScrnAdpt","Unexpected null button");
+            }
+
+            return btn;
+        }
+    }
+
+    public void setButtonVisibility(int androidCode, boolean lookupID, boolean isButtonHidden){
+        int index = getButtonIndex(androidCode, lookupID);
+        boolean toggled = isButtonHidden ^ hiddenButtons[index];
+        hiddenButtons[index] = isButtonHidden;
+        if (visibleButtons != null) {
```

Comment:
If you initialize visibleButtons in the class body, then you don't need to do this null check.